File: src/main/java/shirokuro/embedscript/util/MojangUtil.java
package shirokuro.embedscript.util;

import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import com.google.gson.stream.JsonReader;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

/**
 * @author shirokuro
 */
public final class MojangUtil {
    private static final Cache<UUID, String> NAME_CACHE = CacheBuilder.newBuilder()
        .expireAfterAccess(1, TimeUnit.HOURS)
        .build();
    private static final Cache<String, UUID> UUID_CACHE = CacheBuilder.newBuilder()
        .expireAfterAccess(1, TimeUnit.HOURS)
        .build();

    private MojangUtil() {

    }

    public static String getName(UUID uniqueId) {
        String cached = NAME_CACHE.getIfPresent(uniqueId);
        if (cached != null) {
            return cached;
        }

        String stringUniqueId = uniqueId.toString().replace("-", "");
        String urlString = "https://api.mojang.com/user/profiles/" + stringUniqueId + "/names";
        String name = null;

        try (JsonReader reader = newReader(urlString)) {
            reader.beginArray();
            while (reader.hasNext()) {
                name = readName(reader);
            }
            reader.endArray();
        } catch (IOException e) {
            e.printStackTrace();
        }

        if (name != null) {
            NAME_CACHE.put(uniqueId, name);
            return name;
        }
        return null;
    }

    private static String readName(JsonReader reader) throws IOException {
        String name = null;

        reader.beginObject();
        while (reader.hasNext()) {
            if (reader.nextName().equals("name")) {
                name = reader.nextString();
            } else {
                reader.skipValue();
            }
        }
        reader.endObject();

        return name;
    }

    public static UUID getUUID(String name) {
        UUID cache = UUID_CACHE.getIfPresent(name);
        if (cache != null) {
            return cache;
        }

        String stringUrl = "https://api.mojang.com/users/profiles/minecraft/" + name;
        String stringUniqueId = null;

        try (JsonReader reader = newReader(stringUrl)) {
            reader.beginObject();
            while (reader.hasNext()) {
                if (reader.nextName().equals("id")) {
                    stringUniqueId = reader.nextString();
                } else {
                    reader.skipValue();
                }
            }
            reader.endObject();
        } catch (IOException e) {
            e.printStackTrace();
        }

        if (stringUniqueId != null) {
            UUID uniqueId = toUUID(stringUniqueId);
            UUID_CACHE.put(name, uniqueId);
            return uniqueId;
        }
        return null;
    }

    private static JsonReader newReader(String urlString) throws IOException {
        URL url = new URL(urlString);
        return new JsonReader(new BufferedReader(new InputStreamReader(url.openStream())));
    }

    private static UUID toUUID(String shortUUID) {
        StringBuilder sb = new StringBuilder(36);
        sb.append(shortUUID);
        sb.insert(20, '-');
        sb.insert(16, '-');
        sb.insert(12, '-');
        sb.insert(8, '-');

        return UUID.fromString(sb.toString());
    }
}

File: src/main/kotlin/com/github/kuro46/embedscript/util/MojangUtil.kt
package com.github.kuro46.embedscript.util

import com.google.common.cache.Cache
import com.google.common.cache.CacheBuilder
import com.google.gson.stream.JsonReader
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.URL
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * @author shirokuro
 */
object MojangUtil {
    private val NAME_CACHE: Cache<UUID, String> = CacheBuilder.newBuilder()
        .expireAfterAccess(1, TimeUnit.HOURS)
        .build()
    private val UUID_CACHE: Cache<String, UUID> = CacheBuilder.newBuilder()
        .expireAfterAccess(1, TimeUnit.HOURS)
        .build()

    fun getName(uniqueId: UUID): FindNameResult {
        NAME_CACHE.getIfPresent(uniqueId)?.let {
            return FindNameResult.Found(it)
        }

        val url = uniqueId.toString().replace("-", "")
            .let { "https://api.mojang.com/user/profiles/$it/names" }
        var name: String? = null

        newReader(url).use { reader ->
            reader.beginArray()
            while (reader.hasNext()) {
                name = readName(reader)
            }
            reader.endArray()
        }

        return name?.let {
            NAME_CACHE.put(uniqueId, it)
            FindNameResult.Found(it)
        } ?: FindNameResult.NotFound
    }

    private fun readName(reader: JsonReader): String? {
        var name: String? = null

        reader.beginObject()
        while (reader.hasNext()) {
            if (reader.nextName() == "name") {
                name = reader.nextString()
            } else {
                reader.skipValue()
            }
        }
        reader.endObject()

        return name
    }

    fun getUUID(name: String): FindIdResult {
        UUID_CACHE.getIfPresent(name)?.let {
            return FindIdResult.Found(it)
        }

        val url = "https://api.mojang.com/users/profiles/minecraft/$name"
        var stringUniqueId: String? = null

        newReader(url).use { reader ->
            reader.beginObject()
            while (reader.hasNext()) {
                if (reader.nextName() == "id") {
                    stringUniqueId = reader.nextString()
                } else {
                    reader.skipValue()
                }
            }
            reader.endObject()
        }

        return stringUniqueId?.let {
            val uniqueId = toUUID(it)
            UUID_CACHE.put(name, uniqueId)
            FindIdResult.Found(uniqueId)
        } ?: FindIdResult.NotFound
    }

    private fun newReader(urlString: String): JsonReader {
        val url = URL(urlString)
        return JsonReader(BufferedReader(InputStreamReader(url.openStream(), StandardCharsets.UTF_8)))
    }

    private fun toUUID(shortUUID: String): UUID {
        val sb = StringBuilder(36)
        sb.append(shortUUID)
        sb.insert(20, '-')
        sb.insert(16, '-')
        sb.insert(12, '-')
        sb.insert(8, '-')

        return UUID.fromString(sb.toString())
    }

    sealed class FindIdResult {
        data class Found(val id: UUID) : FindIdResult()
        object NotFound : FindIdResult()
    }

    sealed class FindNameResult {
        data class Found(val name: String) : FindNameResult()
        object NotFound : FindNameResult()
    }
}
